Use lmdb database in picocoin

aido opened the pull request:
Hi,

This pull request is the first step in introducing  a simple and lightweight key-value database to picocoin:
https://symas.com/products/lightning-memory-mapped-database/

LMDB is written in C, extremely high performance and memory-efficient. It is used in other crypto currency projects such as Monero so seems like a good fit for picocoin.
Along with some other handy features LMDB can have multiple sub-databases. I haven't thought things out fully yet but blkdb could possibly be stored as a sub-database with the blockchain as another sub-database or collection of sub-databases, performance permitting.

This would add another building block to make it easier to use libccoin to build a full node if someone so wishes.

aido commented:
Hi,

Given the preference expressed in PR #96 to using normal autoconf toolage for linking to external libraries rather than importing a tree / adding a git subtree into the codebase I assume this PR in its current form is not a good idea.

jgarzik commented:
@aido It's a balance. As a first step we should link externally. If the library proves consensus critical, or versioning through OS proves problematic, we can look at moving in-tree via git subtree or code import.  libevent is stable across most modern OS's.   lmdb might(?) be different.

First step would be to link against external lmdb and change the picocoin code to use lmdb.

aido commented:
OK Jeff,

I'll leave this PR open and play around with it a bit to see if LMDB is a good choice.

aido committed:
ccoin: Add LMDB as a dependency

aido committed:
db: Modularize database code

aido committed:
db: Add lmdb metadb and blockdb databases

aido committed:
db: Replace blocks file with blockdb database

jgarzik commented:
Looking good so far!

aido commented:
Yes, it;s coming along nicely. LMDB seems to be quite suitable as a DB choice.

Next I think I'll add a block height DB to help speed up regenerating blkdb from blockdb on startup (may need to look at the naming of these things). 

I see GitHub has a handy code review feature for each push to a pull request so feel free to add constructive criticism if it is needed.

aido committed:
db: Add a blockheightdb database
This will help speed up regenerating blkdb on startup

aido commented:
Hi, @jgarzik,

The latest commit adds a blockheightdb to speed up rebuilding of blkdb on startup.

Some of the OSX builds are failing on Travis. Travis builds fail if no output is received in 10 minutes. The blockfile test requires ~228,000 block headers to be loaded into the blockheightdb (193,000 mainnet + ~35,000 testnet). This is taking slightly more than 10 minutes on OSX. There is a Trvis workaround for this but it is not ideal and will probably cause the build to fail for other reasons anyway.
The mainnet test headers are in the `test/hdr193000.ser` file. A smaller file containing maybe 100,000 headers should do the trick.

Also, a couple of things I've noticed while testing not really related to thisPR.
`brd` seems to initially download 500 blocks, the `getblocks` limit. It then stops and just listens to the network, If restarted it will then download the next 500 blocks, reading the initial 500 from the blockdb. Is this behaviour by design? If so, without a restart `brd` will only ever relay blocks 0 to 500.
We now have a blockdb and a blkdb, slightly confusing/clashing names. I propose renaming blkdb to chaindb which is probably a more accurate description of what it is. Your thoughts?

aido committed:
test: Reduce number of blkdb tests
Speeds up testing a tad. Mainnet headers reduced from 193000 headers to 50000. Testnet headers reduced from 35141 headers to 25000

aido commented:
Hi,

The purpose of this pull request is to introduce LMDB databases to picocoin. So in that regard this PR is now complete (although it may still need some tidying up).

Travis OSX VMs are really slow at best and sometimes really, really slow. When they are really, really slow some Travis builds timeout, fail and need to be restarted.

I am now looking at using an LMDB database to implement a chaindb/blkdb. But I may do this as a separate PR.

aido committed:
db: Change in memory blkdb database to use data from LMDB databases instead of from file
Also rename blkdb database to chaindb to avoid confusion with LMDB blockdb

aido committed:
brd: Peers list filename now based on chain name

libbitc commented on include/ccoin/db/db.h:
```diff
@@ -0,0 +1,58 @@
+};
+
+ enum {
+ 	MAX_DB_SIZE = 67108864	// Maximum database size in bytes
```
MAX_DB_SIZE should be set to a more realistic valiue e.g.  17179869184

aido committed:
Merge branch 'master' into liblmdb

jgarzik commented:
Taking a fresh look at all the pull requests.

Concept ACK -- lmdb is proving to be superior in other projects, so it makes sense to pull it into picocoin as well.

aido committed:
Merge branch 'master' into liblmdb
